frame/merkle-mountain-range/src/mmr/storage.rs

```diff
@@ -305,43 +334,33 @@ where
 		let mut leaf_index = leaves;
 		let mut node_index = size;
 
-		// Use parent hash of block adding new nodes (this block) as extra identifier
-		// in offchain DB to avoid DB collisions and overwrites in case of forks.
-		let parent_hash = <frame_system::Pallet<T>>::parent_hash();
 		for elem in elems {
-			// For now we store this leaf offchain keyed by `(parent_hash, node_index)`
-			// to make it fork-resistant.
-			// Offchain worker task will "canonicalize" it `frame_system::BlockHashCount` blocks
-			// later when we are not worried about forks anymore (highly unlikely to have a fork
-			// in the chain that deep).
-			// "Canonicalization" in this case means moving this leaf under a new key based
-			// only on the leaf's `node_index`.
-			let key = Pallet::<T, I>::node_offchain_key(parent_hash, node_index);
-			frame_support::log::debug!(
-				target: "runtime::mmr::offchain", "offchain db set: pos {} parent_hash {:?} key {:?}",
-				node_index, parent_hash, key
-			);
-			// Indexing API is used to store the full node content (both leaf and inner).
-			elem.using_encoded(|elem| offchain_index::set(&key, elem));
-
 			// On-chain we are going to only store new peaks.
 			if peaks_to_store.next_if_eq(&node_index).is_some() {
-				<Nodes<T, I>>::insert(node_index, elem.hash());
+				<Peaks<T, I>>::insert(node_index, elem.hash());
 			}
 
 			// Increase the indices.
-			if let Node::Data(..) = elem {
-				leaf_index += 1;
-			}
+			// Store all newly added nodes and leaves to temporary storage,
+			// offchain worker will copy them over to offchain db once block is finished.
 			node_index += 1;
+			match elem {
+				Node::Data(..) => {
+					leaf_index += 1;
+					// TODO: 'try' bounded and throw error instead of truncating.
+					let bounded_encoded_leaf = BoundedVec::<_, _>::truncate_from(elem.encode());
+					LatestLeaf::<T, I>::put(bounded_encoded_leaf);
+				},
+				Node::Hash(..) => <NewNodes<T, I>>::insert(node_index, elem.hash()),
```

Comment:
DQ: is this function (`append`) is only called once per block? If so, then probably it makes sense to change type of `NewNodes` from `StorageMap<NodeIndex => Hash>` to `StorageValue<BoundedVec<(NodeIndex, Hash)>>` to avoid multiple storage writes? Even if it called multiple times, maybe it can be changed as well? I'm not sure - just sharing my thoughts :)

Reply:
This function is called multiple times (at least once) per block: once for adding the new leaf, then for each parent node that has to be created in the MMR to preserve the properties of the structure after adding the new leaf.

I will try to batch these together in some in-mem data struct and commit them together to storage.